### PR TITLE
Fix not being able to upload to a table if the uploads database has switched

### PR DESCRIPTION
**Pull request by calherries**
This PR fixes a bug with uploads on the FE, where switching the uploads database means you can't append to existing uploaded tables anymore.

Steps to reproduce:
1. Set the uploads database to database A
2. Upload a CSV file to database A. 
3. Check that you can append through the newly created model.
3. Set the uploads database to database B
4. You will not be able to see the button to append to the newly created model anymore.

We decided that users should be able to upload to an existing uploaded table after switching the uploads database [here](https://metaboat.slack.com/archives/C04S696LRUM/p1701874496054379).

This is totally possible on the BE but not from the UI.

### CODE REVIEW

File: frontend/src/metabase/query_builder/components/view/ViewHeader/components/QuestionActions/QuestionActions.tsx [hunk 1]
```diff
@@ -94,7 +92,7 @@ export const QuestionActions = ({
   const canWrite = question.canWrite();
   const isSaved = question.isSaved();
   const database = question.database();
-  const canAppend = canUpload && canWrite && !!question._card.based_on_upload;
+  const canAppend = canWrite && !!question._card.based_on_upload;
```
Comment (calherries) [inline frontend/src/metabase/query_builder/components/view/ViewHeader/components/QuestionActions/QuestionActions.tsx:95]: This is the key change. We don't need to check that the question's database has `can_upload = true`. Using based_on_upload` is sufficient to determine whether the user can upload to the table underlying the card.

Here's the docstring for what based_on_upload means:
https://github.com/metabase/metabase/blob/eef0cecb1eaaba5592df70359af09bb3f68903da/src/metabase/upload.clj#L782-L790

We could go a step further than this and do the `canWrite` check on the BE too, so the logic is not split between frontend and backend. Then we could rename `based_on_upload` to `can_append`.

File: frontend/src/metabase/collections/components/CollectionContent/CollectionContent.tsx [hunk 2]
```diff
@@ -47,7 +47,7 @@ export function CollectionContent({
     getSetting(state, "uploads-enabled"),
   );
 
-  const canUploadToDb = useSelector(
+  const canCreateUploadInDb = useSelector(
```
Comment (iethree) [inline frontend/src/metabase/collections/components/CollectionContent/CollectionContent.tsx:50]: I'm curious, what does `canCreateUploadInDb` convey that `canUploadToDb` doesn't?

Comment (calherries) [reply frontend/src/metabase/collections/components/CollectionContent/CollectionContent.tsx:50]: This is my attempt to help prevent the confusion that created the bug in the first place: `database.can_upload = true` actually only means "we can create a new uploaded table", but instead you thought it meant "we can create to a new uploaded table and we can append to an existing table". Making this variable `canCreateUploadInDb` removes that confusion, I think.

**Review by iethree — APPROVED** [review]
Frontend looks good 👍

**Review by crisptrutski — APPROVED** [review]
You didn't ask for a BEC tick, but here it is 😄

File: src/metabase/upload.clj [hunk 3]
```diff
@@ -784,8 +784,10 @@
     - the query is a GUI query, and does not have any joins
     - the base table of the card is based on an upload
     - the user has permissions to upload to the table
-    - uploads are enabled
-  Otherwise based_on_upload is nil."
+    - uploads are enabled for at least one database
```
[the remark this answers is not in the document]
Comment (calherries) [reply src/metabase/upload.clj]: attempted that here [b261b38](https://github.com/metabase/metabase/pull/43026/commits/b261b38dacfa196ea6df2c0e4d41917a417c3794)

**Pull request merged**